### PR TITLE
Feature/character count component

### PR DESCRIPTION
**Pull request by szilvihorvath**
Issue: https://github.com/Capgemini/dcx-react-library/issues/414

Created new CharacterCount component.

Key features:

- Developer should be able to specify the character limit number
- Developer should be able to pass in the character limit prompt message (one when JS is enabled, and another for when non-js is disabled)
- HTML structure should match GDS markup (with ability to set css classes) or the developer should be able to specify the HTML markup
- Character should count down when JS is enabled

### CODE REVIEW

File: src/characterCount/__tests__/CharacterCount.test.tsx [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    await user.type(textarea, 'some text');
+
+    expect(handleChange).toBeCalled();
+    expect(
+      screen.getByText('You can enter up to 15 characters.')
+    ).toBeInTheDocument();
+  });
+});
+
+// it('should not render conditional input field when progressive enhancement is ture but no conditional data', () => {
```
Comment (daniele-zurico) [inline src/characterCount/__tests__/CharacterCount.test.tsx:110]: this test is needed but is commented

**Comment by daniele-zurico** [comment]
Hey @szilvihorvath really well done on your PR. It's almost perfect. I had a look on the gds toolkit documentation and I found out more scenario that we should take in consideration like the following:
[ ] it doesn't stop you writing but change the message when you over the limit (became red)
[ ] it counts the letters (maxlength)
[ ] it counts the words (maxwords)
[ ] it allow a threeshold after that the message appear (maxlength: 112, threshold: 75)

Please can you also keep an eye on the styled version (govUk)... I feel like some classes are not correctly implemented. Probably the class `govuk-textarea` should make the job.

**Review by daniele-zurico — DISMISSED** [review]

File: src/characterCount/CharacterCount.tsx [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+  ...props
+}: CharacterCountProps) => {
+  const [remainingCount, setRemainingCount] = useState<number>(maxLength);
+  const [overLimitBy, setOverLimitBy] = useState<number>(0);
+  const [showMessage, setShowMessage] = useState<boolean>(!threshold);
+  const [showError, setShowError] = useState<boolean>(displayError);
+
+  const isWordsCount = limitType === 'words';
+
+  let hydrated = useHydrated();
```
Comment (Ibabalola) [inline src/characterCount/CharacterCount.tsx:138]:
```suggestion
  const hydrated = useHydrated();
```

File: src/characterCount/CharacterCount.tsx [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+          ])}
+          id={id}
+          name={name}
+          onChange={handleChange}
+          rows={rows}
+          cols={cols}
+          maxLength={constrained && !isWordsCount ? maxLength : undefined}
+          aria-describedby={ariaDescribedBy || ''}
+          {...props}
+        ></textarea>
```
Comment (Ibabalola) [inline src/characterCount/CharacterCount.tsx:198]:
```suggestion
        />
```

**Review by Ibabalola — DISMISSED** [review]

**Review by daniele-zurico — APPROVED** [review]

**Review by daniele-zurico — APPROVED** [review]

**Pull request merged**